### ISSUE TITLE
clang-tidy x86 crashes when modernize-use-nullptr check is enabled.

### ISSUE THREAD

**Issue by kylereedmsft** [llvm/llvm-project#53778]
"C:\Program Files\Microsoft Visual Studio\2022\main\VC\Tools\Llvm\bin\clang-tidy.exe" repro.cpp --checks="modernize-use-nullptr"
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: "C:\\Program Files\\Microsoft Visual Studio\\2022\\main\\VC\\Tools\\Llvm\\bin\\clang-tidy.exe" repro.cpp --checks=modernize-use-nullptr
#0 0x0167f750 C:\Program Files\Microsoft Visual Studio\2022\main\VC\Tools\Llvm\bin\clang-tidy.exe 0xd3f750
0x0167F750 (0x00000069 0x00000104 0x04730CD0 0x0438E038)

LLVM (http://llvm.org/):
  LLVM version 13.0.0
  Optimized build.
  Default target: i686-pc-windows-msvc
  Host CPU: skylake

Host OS: x64 Windows 11.

I'm using the toolset that ships with Visual Studio, but this also repros with the official release.
repro.cpp is an empty file, but it will repro on any file.
This issue does not repro with the x64 version of the tool.
Both modernize-use-nullptr and hicpp-use-nullptr will cause the crash.

**Comment by JonasToth** [comment]
Could you please provide the reproducer here or did I overlook something?

If it contains sensitive code, you can try `creduce`. The reproducer should be self-contained. You can preprocess the file with the `-E` option of clang (i have no clue how to do it with visual studio).

**Comment by kylereedmsft** [comment]
It repros with an empty file. It seems to be a problem if the check is even enabled.

**Comment by kylereedmsft** [comment]
Yes, it's only this check. I started with * and then narrowed it down from there.
It only reproduces when using the x86 build of the tools on x64 Windows.
I just tried again with 13.0.1 and it's still failing there.

**Comment by kylereedmsft** [comment]
It does not repro in 11.0.0

**Comment by njames93** [comment]
> Yes, it's only this check. I started with * and then narrowed it down from there.
> It only reproduces when using the x86 build of the tools on x64 Windows.
> I just tried again with 13.0.1 and it's still failing there.

Unrelated but that pain should hopefully be alleviated in clang-tidy-15 😃

**Comment by fredriksd** [comment]
@EugeneZelenko, @njames93, we have run into the same issue when running clang-tidy on x86-windows. It fails with 14.0.6 and 15.0.5 both.

Is there any update from you guys on if this would be fixed at some point?

**Comment by njames93** [comment]
Can you please provide a stack trace of the crash when running clang-tidy-15

**Comment by fredriksd** [comment]
> Can you please provide a stack trace of the crash when running clang-tidy-15
Attaching an output of a reproduced case. `main.cpp` is completely empty.

```
c:\dev\github\NordicPlayground\nrf-device-lib>clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 15.0.5
  Optimized build.
  Default target: i686-pc-windows-msvc
  Host CPU: skylake

c:\dev\github\NordicPlayground\nrf-device-lib>clang-tidy main.cpp --checks=modernize-use-nullptr
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "main.cpp"
No compilation database found in c:\dev\github\NordicPlayground\nrf-device-lib or any parent directory
fixed-compilation-database: Error while opening fixed database: no such file or directory
json-compilation-database: Error while opening JSON database: no such file or directory
Running without flags.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: clang-tidy main.cpp --checks=modernize-use-nullptr
Exception Code: 0xC0000005
#0 0x01769780 C:\Program Files (x86)\LLVM\bin\clang-tidy.exe 0x7c9780
0x01769780, C:\Program Files (x86)\LLVM\bin\clang-tidy.exe(0x0000000000FA0000) + 0x7C9780 byte(s)
```

**Comment by njames93** [comment]
Is that the whole stack trace?

**Comment by fredriksd** [comment]
That's all the information I get out from it, yes. I'm not sure if there is any way to get more verbose output?

**Comment by Anthony-Nicholls** [comment]
I just wanted to add that I've only been able to reproduce this issue with the version bundled with VS2022. I created an example on GitHub Actions and added jobs for the following
- llvm version 15.0.4 (installed using choco)
- llvm version 15.0.5 (installed using choco)
- llvm version 15.0.7 (installed using choco)
- llvm version 16.0.0 (installed using choco)
- llvm version 16.0.1 (installed using choco)
- llvm version 16.0.2 (installed using choco)
- llvm version 16.0.3 (installed using choco)
- llvm version 16.0.4 (installed using choco)
- llvm version 16.0.5 (installed using choco)
- llvm version 16.0.6 (installed using choco)
- llvm version bundled with windows-2022 image on GitHub Actions
- llvm version bundled with Visual Studio Enterprise 2022 v17.6.33829.357 as installed on the windows-2022 image on GitHub Actions
  - note clang-tidy reports version 15.0.1 but I've also had the exact same job report 16.0.5!?
- llvm version bundled with Visual Studio Enterprise 2019 v16.11.33927.289 as installed on the windows-2019 image on GitHub Actions
- llvm version bundled with the ubuntu-22.04 runner on GitHub Actions

You can access the repository and results [here](https://github.com/Anthony-Nicholls/clang-tidy-crash-example/actions/runs/5973632848/job/16206259185).

**Comment by eljonny** [comment]
I'm experiencing this issue with VS 2022 17.13.0.
In the VS PowerShell dev console with this header:

```
**********************************************************************
** Visual Studio 2022 Developer PowerShell v17.13.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
```

Clang-Tidy version 19.1.1:

```
PS C:\Users\user\Code\TestCPP> clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 19.1.1
  Optimized build.
```

This is the message I get, which is basically the same as the previously-posted message:

```
PS C:\Users\user\Code\TestCPP> clang-tidy src\TestCPPAssertions.cpp                                                    Error while trying to load a compilation database:
Could not auto-detect compilation database for file "src\TestCPPAssertions.cpp"
No compilation database found in C:\Users\jonat\Code\TestCPP\src or any parent directory
fixed-compilation-database: Error while opening fixed database: no such file or directory
json-compilation-database: Error while opening JSON database: no such file or directory
Running without flags.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\Llvm\\bin\\clang-tidy.exe" src\\TestCPPAssertions.cpp
Exception Code: 0xC0000005
#0 0x011e0ca0 (C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\bin\clang-tidy.exe+0x770ca0)
```

Then when I disable the checks mentioned in this issue, it works properly.
Per [this VS Developer Feedback Issue filed on 2025-01-02](https://developercommunity.visualstudio.com/t/clang-tidy-terminates-with-Exception-Cod/10822069?pageSize=15&sort=active&openOnly=false&closedOnly=false&topics=repos) MS is saying this is an LLVM/Clang issue with clang-tidy, and referencing this issue.
I am using the LLVM version that gets installed with Visual Studio 2022.
Just wanted to add to this to let y'all know that this is still an issue in 19.1.1.

**Comment by eljonny** [comment]
Is anyone looking at this issue?
I've never worked in any LLVM source before but I would consider looking at this to fix it if nobody is looking at it.

**Comment by EugeneZelenko** [comment]
Will be good idea to try 20 or `main` branch first.

**Comment by eljonny** [comment]
I'll check that out first. Thank you Eugene.

**Comment by EugeneZelenko** [comment]
@eljonny: It looks like reproducer is missing, so you'll need to create one. If you'll have questions about Clang-Tidy code, @vbvictor and @HerrCai0907 (who are quite active currently) could help you.

**Comment by eljonny** [comment]
Thanks so much for the information @EugeneZelenko! I'll get this rolling if and after I confirm it's still an issue.
